clm/model_utils_gpu.py

```diff
@@ -0,0 +1,152 @@
+    return dict(items)
+
+def _modify_config(gpt_cfg, cfg, add_cfg_to_tree=False):
+    """
+    This function modifies the original gpt pre-training config (gpt_cfg) with attributes from the finetuning config (cfg).
+    The `add_cfg_to_tree` arg adds `cfg` to the top of the yaml tree which is needed for all `hparams.yaml` files when passed as an arg to `load_from_checkpoint()`.
+    """
+    OmegaConf.set_struct(gpt_cfg, True)
+    OmegaConf.resolve(cfg)
+    with open_dict(gpt_cfg):
+        gpt_cfg.megatron_amp_O2 = cfg.model.get("megatron_amp_O2", False)
+        gpt_cfg.micro_batch_size = cfg.model.data.train_ds.micro_batch_size
+        gpt_cfg.global_batch_size = cfg.model.data.train_ds.global_batch_size
+        gpt_cfg.sequence_parallel = cfg.model.get("sequence_parallel", False)
+        gpt_cfg.activations_checkpoint_granularity = cfg.model.get("activations_checkpoint_granularity", None)
+        gpt_cfg.activations_checkpoint_num_layers = cfg.model.get("activations_checkpoint_num_layers", None)
+        gpt_cfg.activations_checkpoint_method = cfg.model.get("activations_checkpoint_method", None)
+        gpt_cfg.activations_checkpoint_layers_per_pipeline = cfg.model.get(
+            "activations_checkpoint_layers_per_pipeline", None
+        )
+        gpt_cfg.peft = cfg.model.peft
+        gpt_cfg.data = cfg.model.data
+        gpt_cfg.optim = cfg.model.optim
+        gpt_cfg.precision = cfg.trainer.precision
+        gpt_cfg.answer_only_loss = cfg.model.answer_only_loss
+        gpt_cfg.restore_from_path = cfg.model.restore_from_path
+        gpt_cfg.resume_from_checkpoint = cfg.model.resume_from_checkpoint
+        gpt_cfg.save_nemo_on_validation_end = cfg.model.save_nemo_on_validation_end
+        gpt_cfg.gradient_as_bucket_view = cfg.model.gradient_as_bucket_view
+        gpt_cfg.hidden_dropout = cfg.model.get("hidden_dropout", 0.0)
+        gpt_cfg.attention_dropout = cfg.model.get("attention_dropout", 0.0)
+        gpt_cfg.ffn_dropout = cfg.model.ffn_dropout
+        gpt_cfg.use_flash_attention = cfg.model.get("use_flash_attention", False)
+        # if TP/PP size is -1, use default TP/PP size as original model
+        if cfg.model.get("tensor_model_parallel_size", 1) > 0:
+            gpt_cfg.tensor_model_parallel_size = cfg.model.get("tensor_model_parallel_size", 1)
+        if cfg.model.get("pipeline_model_parallel_size", 1) > 0:
+            gpt_cfg.pipeline_model_parallel_size = cfg.model.get("pipeline_model_parallel_size", 1)
+        gpt_cfg.pipeline_model_parallel_split_rank = cfg.model.get("pipeline_model_parallel_split_rank", 0)
+
+        if cfg.model.data.get("chat", False):
+            # chat model, overwrite the prompt template
+            prompt_template = get_prompt_template_example(cfg.model.data.chat_prompt_tokens)
+            gpt_cfg.data.train_ds.prompt_template = prompt_template
+            gpt_cfg.data.validation_ds.prompt_template = prompt_template
+
+        sft_cls = GPTSFTModel
```

Comment:
At first glance, looks like `GPTSFTModel` is a class in nemo aligner which covers `NLPAdapterModelMixin` specifically [here](https://github.com/NVIDIA/NeMo-Aligner/blob/d8ef9fb05dc9a30780d5f0092b918476fecd83be/nemo_aligner/models/nlp/gpt/gpt_sft_model.py#L49)

Given no need to cover PEFT adapter, we can directly use `MegatronGPTModel` instead if possible.
* [training `MegatronGPTModel`](https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/megatron_gpt_pretraining.py)
* [mlperf example `MegatronGPTModel`](https://github.com/mlcommons/training_results_v3.1/blob/main/NVIDIA/benchmarks/gpt3/implementations/pytorch/custom_callbacks.py#L523)

If we we need some api in `SupervisedInterface` feel free to leave as it is.